framework/src/main/java/org/checkerframework/common/basetype/SideEffectsOnlyAnnoChecker.java

```diff
@@ -0,0 +1,198 @@
+package org.checkerframework.common.basetype;
+
+import com.sun.source.tree.AssignmentTree;
+import com.sun.source.tree.CatchTree;
+import com.sun.source.tree.CompoundAssignmentTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.NewClassTree;
+import com.sun.source.tree.Tree;
+import com.sun.source.tree.UnaryTree;
+import com.sun.source.util.TreePath;
+import com.sun.source.util.TreePathScanner;
+import java.util.ArrayList;
+import java.util.List;
+import javax.annotation.processing.ProcessingEnvironment;
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.Element;
+import javax.lang.model.element.ExecutableElement;
+import org.checkerframework.dataflow.expression.JavaExpression;
+import org.checkerframework.dataflow.qual.Pure;
+import org.checkerframework.dataflow.qual.SideEffectFree;
+import org.checkerframework.dataflow.qual.SideEffectsOnly;
+import org.checkerframework.framework.util.JavaExpressionParseUtil;
+import org.checkerframework.framework.util.StringToJavaExpression;
+import org.checkerframework.javacutil.AnnotationProvider;
+import org.checkerframework.javacutil.AnnotationUtils;
+import org.checkerframework.javacutil.Pair;
+import org.checkerframework.javacutil.TreeUtils;
+
+/**
+ * For methods annotated with {@link SideEffectsOnly}, computes expressions that are side-effected
+ * but not permitted by the annotation.
+ */
+public class SideEffectsOnlyAnnoChecker {
```

Comment:
I am not aware of a mutation checker and couldn’t find it in the manual either. Could you please point me to that class?
Although I think the `PurityChecker` can be modified to add all the functionality in the `SideEffectsOnlyAnnoChecker`. Do you think that's more appropriate?

Reply:
By "mutation checker" I meant `PurityChecker`.  I'm sorry for being sloppy in my terminology.  Thank you for setting me straight.  Yes, I was thinking that merging the functionality in one place (maybe controlled by a switch) would be shorter, easier to understand, and less prone to code drift, compared to the two separate implementations.  At least, I think that is worth trying to see whether it works out.

Thank you!

Reply:
Thank you for the suggestions. I tried merging the two functionalities into `PurityChecker` but wasn’t successful because of the following reason:
`SideEffectsOnlyAnnoChecker` (a part of the Checker Framework repository) needs to import `org.checkerframework.common.basetype.BaseTypeChecker`, which is not visible in the dataflow project (dataflow project contains the `PurityChecker`). Is there any other way around this?